### PR TITLE
Simplify `TokenTree` and fix `macro_rules!` bugs

### LINKED ISSUE #39390: macros: `tt` fragments treat `$(foo)*` like `foo`
For example, the following compiles and prints `success`:
```rust
macro_rules! identity {
    ($t:tt) => { $t }
}

identity! {
    fn $(main)* () {
        $($( println!($("success")+); )+)+
    }
}
```

### LINKED ISSUE #39403: macros: simplify `tt` matchers
Ideally, `tt` matches either a single token or delimited sequence (e.g. `{ arbitrary tokens }`).
However, the following compiles today:
```rust
macro_rules! m { ($t:tt) => {} }
m!($x);
m!($(x)*);
m!($x:ident);
```
We want to avoid leaking details of `macro_rules!` syntax into the semantics of `tt`, at least for macros 2.0.

### LINKED ISSUE #39404: macros: syntax errors in matches should fail on definition, not use
For example,
```rust
macro_rules! m { ($i) => {} }
// m!(); // uncommenting this causes an error at the def site
```

### LINKED ISSUE #39889: Regression in procedural macros emitting macro_rules macros
#### `my-macro/src/lib.rs`

```rust
extern crate proc_macro;
use proc_macro::TokenStream;

#[proc_macro_derive(Useless)]
pub fn demo(_input: TokenStream) -> TokenStream {
    let rules = r#"
        macro_rules! id {
            ($($tt:tt)*) => { $($tt)* };
        }
    "#;
    rules.parse().unwrap()
}
```

#### `testing/src/main.rs`

```rust
#[macro_use]
extern crate my_macro;

#[derive(Useless)]
struct _S;

pub fn main() {
    println!("{}", id!(true));
}
```

#### `rustup run nightly-2017-02-13 cargo run`

```
true
```

#### `rustup run nightly-2017-02-16 cargo run`

```
error: attempted to repeat an expression containing no syntax variables matched as repeating at this depth
 --> src/main.rs:4:10
  |
4 | #[derive(Useless)]
  |          ^^^^^^^

error: attempted to repeat an expression containing no syntax variables matched as repeating at this depth
 --> src/main.rs:4:10
  |
4 | #[derive(Useless)]
  |          ^^^^^^^

error: aborting due to 2 previous errors
```

cc @jseyfried who touched macro expansion recently

### PR DESCRIPTION
**Pull request by jseyfried**
This PR
 - fixes #39390, fixes #39403, and fixes #39404 (each is a [breaking-change], see issues for examples),
 - fixes #39889,
 - simplifies and optimizes macro invocation parsing,
 - cleans up `ext::tt::transcribe`,
 - removes `tokenstream::TokenTree::Sequence` and `Token::MatchNt`,
   - instead, adds a new type `ext::tt::quoted::TokenTree` for use by `macro_rules!` (`ext::tt`)
 - removes `parser.quote_depth` and `parser.parsing_token_tree`, and
 - removes `quote_matcher!`.
   - Instead, use `quote_tokens!` and `ext::tt::quoted::parse` the result with `expect_matchers=true`.
   - I found no outside uses of `quote_matcher!` when searching Rust code on Github.

r? @nrc

### CODE REVIEW

**Comment by jseyfried** [comment]
This is groundwork for both procedural and declarative macros 2.0.
cc #38356 #39412

File: src/libsyntax/ext/tt/quoted.rs [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, RustcEncodable, RustcDecodable, Hash)]
+pub enum TokenTree {
+    Token(Span, token::Token),
+    Delimited(Span, Rc<Delimited>),
+    /// A kleene-style repetition sequence with a span
+    Sequence(Span, Rc<SequenceRepetition>),
+    /// Matches a nonterminal. This is only used in the left hand side of MBE macros.
+    Match(Span, ast::Ident /* name to bind */, ast::Ident /* kind of nonterminal */),
```
Comment (eddyb) [inline src/libsyntax/ext/tt/quoted.rs]: Can't this be done on the fly? (i.e in parsing not lexing)

Comment (jseyfried) [inline src/libsyntax/ext/tt/quoted.rs]: Neither parsing nor lexing knows about sequences and matches at all. We lex directly into a `Vec<TokenTree>` and consume the token trees in the parser directly (e.g. parsing a token tree is constant time). Sequences and matches are parsed by `ext::tt::quoted::parse`.

File: src/libsyntax/ext/tt/transcribe.rs [hunk 2]
```diff
@@ -14,27 +14,70 @@ use errors::Handler;
...
 use std::collections::HashMap;
 
-///an unzipping of `TokenTree`s
-#[derive(Clone)]
-struct TtFrame {
-    forest: TokenTree,
-    idx: usize,
-    dotdotdoted: bool,
-    sep: Option<Token>,
+enum Frame {
```
Comment (nrc) [inline src/libsyntax/ext/tt/transcribe.rs]: This would benefit from a comment

File: src/libsyntax/ext/tt/transcribe.rs [hunk 3]
```diff
@@ -74,143 +74,46 @@ impl Iterator for Frame {
...
-                    MatchedNonterminal(_) => LisUnconstrained,
-                    MatchedSeq(ref ads, _) => LisConstraint(ads.len(), name),
-                },
-                _ => LisUnconstrained
-            },
-        TokenTree::Token(..) => LisUnconstrained,
-    }
-}
+    let mut stack = SmallVector::one(Frame::Delimited {
+        forest: Rc::new(tokenstream::Delimited { delim: token::NoDelim, tts: src }),
```
Comment (nrc) [inline src/libsyntax/ext/tt/transcribe.rs]: I would factor this out into a ctor function

File: src/libsyntax/ext/tt/transcribe.rs [hunk 4]
```diff
@@ -219,73 +122,142 @@ fn tt_next_token(r: &mut TtReader, prev_span: Span) -> Option<TokenTree> {
...
+            }
+        })
+    })
+}
+
+#[derive(Clone)]
+enum LockstepIterSize {
+    LisUnconstrained,
+    LisConstraint(usize, Ident),
+    LisContradiction(String),
```
Comment (nrc) [inline src/libsyntax/ext/tt/transcribe.rs]: nit: I'd kill the `Lis` prefix

File: src/libsyntax/ext/tt/quoted.rs [hunk 5]
```diff
@@ -0,0 +1,228 @@
+// Copyright 2017 The Rust Project Developers. See the COPYRIGHT
```
Comment (nrc) [inline src/libsyntax/ext/tt/quoted.rs]: Why make this module here? I'd have thought TTs were much more general than just quasi quoting

**Review by nrc — COMMENTED** [review]
I only got halfway through reviewing this PR before the (long) weekend, and then my browser crashed and lost my comments. There wasn't anything major. Anyway, here is a single comment and I'll finish the review on Tuesday. Sorry for the delay.

File: src/libsyntax/ext/tt/transcribe.rs (hunk 2 above)
Comment (jseyfried) [inline src/libsyntax/ext/tt/transcribe.rs]: Will do

File: src/libsyntax/ext/tt/transcribe.rs (hunk 3 above)
Comment (jseyfried) [inline src/libsyntax/ext/tt/transcribe.rs]: Will do

File: src/libsyntax/ext/tt/transcribe.rs (hunk 4 above)
Comment (jseyfried) [inline src/libsyntax/ext/tt/transcribe.rs]: Will do

File: src/libsyntax/ext/tt/quoted.rs (hunk 5 above)
Comment (jseyfried) [inline src/libsyntax/ext/tt/quoted.rs]: We still have `tokenstream::{TokenStream, TokenTree, ...}`.

`quoted::TokenTree` is more general than `tokenstream::TokenTree` -- it also supports "`$`-quoted tokens" from the `macro_rules!` syntax -- `$e`, `$e:expr`, and `$(...)*`. In other words, `quoted::TokenTree` is the parse tree for `macro_rules!` syntax.

**Comment by nrc** [comment]
Oh, GH didn't lose my earlier comments. Good.

**Review by nrc — APPROVED** [review]
Just one more comment, and that can be done later

File: src/libsyntax/ext/tt/quoted.rs [hunk 6]
```diff
@@ -76,6 +76,8 @@ pub enum TokenTree {
     Delimited(Span, Rc<Delimited>),
     /// A kleene-style repetition sequence with a span
     Sequence(Span, Rc<SequenceRepetition>),
+    /// Matches a nonterminal. This is only used in the left hand side of MBE macros.
+    Match(Span, ast::Ident /* name to bind */, ast::Ident /* kind of nonterminal */),
```
Comment (jseyfried) [inline src/libsyntax/ext/tt/quoted.rs]: With a bit more refactoring I could also move `Nonterminal::SubstNt` here -- I think `MacroVariable` would be a better fit for that. `MacroVariableDeclaration` is too long though; how about `MetaVarDecl`?

Comment (nrc) [inline src/libsyntax/ext/tt/quoted.rs]: `MetaVarDecl` SGTM

**Comment by jseyfried** [comment]
@eddyb would you mind Cratering this?

**Comment by eddyb** [comment]
@jseyfried Can you ping me after a rebase?

**Comment by jseyfried** [comment]
@eddyb rebased.

**Comment by eddyb** [comment]
The [crater report](https://gist.github.com/eddyb/fc2733f374e4f1277691062fab8942ad) shows 4 main regressions:
* [ ] [`brotli`](https://crates.io/crates/brotli) ([failure](https://taskcluster.github.io/unified-logviewer/?url=https%3A%2F%2Fqueue.taskcluster.net%2Fv1%2Ftask%2FZmBQRxgsTAS-Y3NK8MoW6w%2Fruns%2F0%2Fartifacts%2Fpublic%2Flogs%2Flive.log&highlightStart=86&highlightEnd=86&lineNumber=&wrapLines=false&showLineNumbers=true&jumpToHighlight=true&followLog=false&lineHeight=19&asText=false))
* [ ] [`chipmunk`](https://crates.io/crates/chipmunk) ([failure](https://taskcluster.github.io/unified-logviewer/?url=https%3A%2F%2Fqueue.taskcluster.net%2Fv1%2Ftask%2FESLkjp4eRx-a0KiqTR2mFQ%2Fruns%2F0%2Fartifacts%2Fpublic%2Flogs%2Flive.log&highlightStart=94&highlightEnd=94&lineNumber=&wrapLines=false&showLineNumbers=true&jumpToHighlight=true&followLog=false&lineHeight=19&asText=false)) - this one looks like a bug in `chipmunk` itself
* [ ] [`clap`](https://crates.io/crates/clap) ([failure](https://taskcluster.github.io/unified-logviewer/?url=https%3A%2F%2Fqueue.taskcluster.net%2Fv1%2Ftask%2F3SSZW_ayTKSBU5URC3Kong%2Fruns%2F0%2Fartifacts%2Fpublic%2Flogs%2Flive.log&highlightStart=136&highlightEnd=136&lineNumber=&wrapLines=false&showLineNumbers=true&jumpToHighlight=true&followLog=false&lineHeight=19&asText=false)) - this is *very bad*, as `clap` has 454 other packages depending on it
* [ ] [`multipart`](https://crates.io/crates/multipart) ([failure](https://taskcluster.github.io/unified-logviewer/?url=https%3A%2F%2Fqueue.taskcluster.net%2Fv1%2Ftask%2FGlgAEw2WQKuUreVF1DTMNw%2Fruns%2F0%2Fartifacts%2Fpublic%2Flogs%2Flive.log&highlightStart=240&highlightEnd=240&lineNumber=&wrapLines=false&showLineNumbers=true&jumpToHighlight=true&followLog=false&lineHeight=19&asText=false))

These all look like macro-defining macros, and they seem like they should work, but something is perhaps doing too-eager parsing?

**Comment by jseyfried** [comment]
It looks like all these errors are due to fixing #39404. That is, they are legitimate errors in the macro definition that didn't show up before because the erroneous pattern never ended up getting used to match anything. I'll add a warning cycle.

> I've never liked `$foo:bar` (or even just `$foo`) being a single token, it shouldn't.

Agreed, that's one of the reasons for this PR. After this PR, `$foo:bar` is never a single token (it used to be `Token::Interpolated(Nonterminal::MatchNt(..))` sometimes; this PR removes `Nonterminal::MatchNt`).
Also, `$foo` is never a single token in a `tokenstream::TokenTree` after this PR, except on certain error paths (planning on removing `Nonterminal::SubstNt` later after some more refactoring).

**Comment by eddyb** [comment]
Is it really #39404? It looked like the lone `$foo` was supposed to be interpolated from an outer macro, or am I imagining things? The fallout looks pretty bad if it's their bug, multiple patch releases may be needed.

**Comment by eddyb** [comment]
Okay, this might not be that bad. `clap` is affected in `1.*` and `2.*` and since these are post-`0.*`, one release for each (of `1.*` and `2.*`) should be enough to fix most (ideally, all) of the regressions.
I've opened kbknapp/clap-rs#857 and we can do another crater run once that's released.

**Comment by jseyfried** [comment]
@SimonSapin
> Regarding hygiene, `FromStr for TokenStream` is already stable, so it will still need to be supported when Macros 2.0 are fully implemented

True, but the scope/hygiene for local variables from `impl FromStr for TokenStream` is not fully stable. `impl FromStr for TokenStream` can crate local variables in an item on stable, but the local variables in scope at the invocation site are always blocked by the item, so they are not yet relevant.

@eddyb
> `#[derive]` has the amazing property of not having any hygienic inputs

Sadly, this isn't exactly true. `$crate` [is hygienic](https://github.com/rust-lang/rust/pull/37463), and this can cause problems today. For example,
```rust
#![feature(use_extern_macros)] // or `#![feature(proc_macro)]`, which implies this

// crate A:
pub struct Z;

#[macro_export]
macro_rules! m { () => {
    #[derive(Custom)]
    struct Y($crate::Z);
} }

// crate B:
extern crate A;
pub use A::m;

// crate C:
extern crate B;
B::m!();
//^ This expands to `#[derive(Custom)] struct Y($crate::Z);`.
//| We can't stringify `$crate::Z` to resolve correctly to `Z` from crate `A`.
```
Today, we make an effort (not yet best effort) to stringify correctly.

@SimonSapin
> so what does that say about the behavior that `FromStr for TokenStream` should have in a `proc_macro` when everything will be implemented

Not much -- I think there are a couple of reasonable points in the design space. For example, we could emit an ambiguity error if there is more than once correct choice for hygiene of a `FromStr for TokenStream`-generated variable.

@akiselev
> Does the Rustlang team have a ballpark guess on when a Macros 2.0 implementation will be started, let alone hit nightly?

Macros 2.0 implementation is under way (see the roadmaps for [macro modularization](https://github.com/rust-lang/rust/issues/35896#issuecomment-277870744), [procedural macros](https://github.com/rust-lang/rust/issues/38356#issuecomment-274377210), and [declarative macros](https://github.com/rust-lang/rust/issues/39412#issuecomment-277867338). More `TokenStream` API should start landing by mid-March.

@SimonSapin
> I think it’s possible to hack something with derive in item position (https://github.com/rust-lang/rfcs/pull/1913#issuecomment-281462534)

That doesn't work in general since attributes do not allow arbitrary tokens. However, this should work on nightly once I support arbitrary tokens in attributes next week, and we could fast-track that to stable.

@plietar
> Should the macro get information on where it is being used?

Not as an argument to the `#[proc_macro]` function (imo), but perhaps via the "macro context" (thread-local state). Regardless, we can add new `#[proc_macro]` function signatures backwards-compatibly later.

@Ericson2314
> Last I talked to @jseyfried all this [hygiene design] was going well

Indeed, although I've been prioritizing unrelated groundwork for procedural macros 2.0. That being said, the hygiene stuff should hit nightly via declarative macros 2.0 sometime in March.

**Comment by jseyfried** [comment]
I added a warning cycle -- r? @nrc

**Comment by nrc** [comment]
r = me, but needs a rebase

**Comment by alexcrichton** [comment]
@bors: retry

* network error

**Pull request merged**